cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/FieldLineageDataset.java

```diff
@@ -0,0 +1,254 @@
+   * Operations represent transformations from source endpoints to the destination endpoints.
+   * From source perspective the operations are added as forward lineage, while from
+   * destination perspective they are added as backward lineage.
+   *
+   * @param programRunId program run for which lineage is to be added
+   * @param info the FieldLineageInfo created by program run
+   */
+  public void addFieldLineageInfoReferenceRecords(ProgramRunId programRunId, FieldLineageInfo info) {
+    // For all the sources, operations represents the Forward lineage
+    for (EndPoint source : info.getSources()) {
+      addOperationReferenceRecord(programRunId, source, info.getChecksum(), FORWARD_LINEAGE_MARKER);
+    }
+
+    // For all the destinations, operations represents backward lineage
+    for (EndPoint destination : info.getDestinations()) {
+      addOperationReferenceRecord(programRunId, destination, info.getChecksum(), BACKWARD_LINEAGE_MARKER);
+    }
+  }
+
+  /**
+   * Get the field lineage information for a given endpoint within the given time range.
+   * TODO: Read methods are mainly for the testing purpose till representation from UI is finalized.
+   *
+   * @param endPoint the endpoint for which lineage information to be retrieved
+   * @param start start time in seconds
+   * @param end end time in seconds
+   * @return {@link Set} of {@link FieldLineageInfo} instances
+   */
+  public Set<FieldLineageInfo> getFieldLineageInfosInRange(EndPoint endPoint, long start, long end) {
+    Set<Long> checksums = getChecksumsInRange(getScanStartKey(endPoint, end, BACKWARD_LINEAGE_MARKER),
+                                              getScanEndKey(endPoint, start, BACKWARD_LINEAGE_MARKER));
+
+    Set<FieldLineageInfo> infos = new HashSet<>();
+    for (long checksum : checksums) {
+      infos.add(getFieldLineageInfo(checksum));
+    }
+    return infos;
+  }
+
+  private FieldLineageInfo getFieldLineageInfo(long checksum) {
+    byte[] rowKey = getChecksumRowKey(checksum);
+    byte[] bytes = table.get(rowKey, RAW_FORMAT_COLS_BYTE);
+    return GSON.fromJson(Bytes.toString(bytes), FieldLineageInfo.class);
+  }
+
+  private byte[] getScanStartKey(EndPoint endPoint, long end, char lineageDirectionMarker) {
+    // time is inverted, hence we need to have end time in start key.
+    return getScanKey(endPoint, end, lineageDirectionMarker);
+  }
+
+  private byte[] getScanEndKey(EndPoint endPoint, long start, char lineageDirectionMarker) {
+    // time is inverted, hence we need to have start time in end key.
+    return getScanKey(endPoint, start, lineageDirectionMarker);
+  }
+
+  private Set<Long> getChecksumsInRange(byte[] startKey, byte[] endKey) {
+    Set<Long> checksums = new HashSet<>();
+    try (Scanner scanner = table.scan(startKey, endKey)) {
+      Row row;
+      while ((row = scanner.next()) != null) {
+        if (LOG.isTraceEnabled()) {
+          LOG.trace("Got row key = {}", Bytes.toString(row.getRow()));
```

Comment:
If you trace log this, then why not include all the useful info? (The run id, the timestamp and the checksum). Also would be good to be able to associate this trace message with the original request, so including the endpoint would help, too).